chore: Changes the DatabaseSelector and TableSelector to use the new Select component

michael-s-molina opened the pull request:
### SUMMARY
Changes the `DatabaseSelector` and `TableSelector` to use the new Select component
- Make the selects more similar to the rest of the application with the labels on top and no tags
- Adds pagination and lazy loading to the database select
- Adds lazy loading to the schemas select
- Adds lazy loading to the tables select and sort the options
- Removes the number of items in a select due to lazy loading
- Improves the layout of the virtual dataset
- Migrates the table and view icons to the new Icons component
- Adjusts the size of the table, view, certification, and warning icons
- Cancels the import of icons when unmounting
- Adjusts the tests

@junlincc @rusackas @villebro @jinghua-qa @rosemarie-chiu

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
https://user-images.githubusercontent.com/70410625/129966548-62b6efa8-44a5-43fa-b9d1-8a59f13b9024.mov

https://user-images.githubusercontent.com/70410625/129966741-03fb26fb-4b53-45eb-881d-d2a115953b86.mov

PS: There's an error that is also happening on master when we try to change the table of a saved dataset. It raises "An error has occurred". On the server logs, we can see a 401. @villebro This is not related to this PR but if it's something simple we can add the fix here.

### TESTING INSTRUCTIONS
Check the videos for instructions.

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

michael-s-molina committed:
chore: Changes the DatabaseSelector to use the new Select component

villebro commented:
> PS: There's an error that is also happening on master when we try to change the table of a saved dataset. It raises "An error has occurred". On the server logs, we can see a 401. @villebro This is not related to this PR but if it's something simple we can add the fix here.

This now fixed on master as of #16369

villebro commented on superset-frontend/src/components/DatabaseSelector/index.tsx:
```diff
@@ -97,193 +89,189 @@ export default function DatabaseSelector({
+        const queryParams = rison.encode({ force: refresh > 0 });
+        const endpoint = `/api/v1/database/${currentDb.value}/schemas/?q=${queryParams}`;
+
+        // The endpoint currently does not support pagination
+        return SupersetClient.get({ endpoint }).then(({ json }) => {
```
Could potentially change this to a `TODO:` if you have to push commits to the PR (would be nice to add pagination in a follow-up)

villebro commented on superset-frontend/src/components/TableSelector/index.tsx:
```diff
@@ -129,179 +175,173 @@ const TableSelector: FunctionComponent<TableSelectorProps> = ({
+        const encodedSchema = encodeURIComponent(dbSchema);
+        const encodedSubstr = encodeURIComponent(search || 'undefined');
+        const endpoint = encodeURI(
+          `/superset/tables/${currentDbId}/${encodedSchema}/${encodedSubstr}/true/`,
```
This should also default to not force updating the metadata from the db
```suggestion
          `/superset/tables/${currentDbId}/${encodedSchema}/${encodedSubstr}/false/`,
```

villebro commented on superset-frontend/src/components/TableSelector/index.tsx:
```diff
@@ -129,179 +175,173 @@ const TableSelector: FunctionComponent<TableSelectorProps> = ({
 
-  function changeSchema(schemaOpt: any, force = false) {
-    const value = schemaOpt ? schemaOpt.value : null;
+  function onRefresh() {
```
I think we need to add a force param here that the force refresh calls. Something like
```js
  function onRefresh(forceRefresh: boolean) {
```

villebro commented on superset-frontend/src/components/TableSelector/index.tsx:
```diff
@@ -129,179 +175,173 @@ const TableSelector: FunctionComponent<TableSelectorProps> = ({
+  const [refresh, setRefresh] = useState(0);
+
+  const loadTable = useMemo(
+    () => async (dbId: number, schema: string, tableName: string) => {
       const endpoint = encodeURI(
-        `/superset/tables/${actualDbId}/${encodedSchema}/${encodedSubstr}/${!!forceRefresh}/`,
+        `/superset/tables/${dbId}/${schema}/${encodeURIComponent(
+          tableName,
+        )}/true/true`,
       );
```
This will force table metadata to be refetched from the db as opposed to the cache. We should probably retain the `forceRefresh` variable and pass it along here, like
```typescript
  const loadTable = useMemo(
    () => async (dbId: number, schema: string, tableName: string, forceRefresh: boolean) => {
      const endpoint = encodeURI(
        `/superset/tables/${dbId}/${schema}/${encodeURIComponent(
          tableName,
        )}/${forceRefresh}/true`,
      );```

villebro reviewed (changes requested):
Great improvement! One last problem related to always force refreshing the data from the db, other than that looks great!

michael-s-molina commented:
@villebro I changed the `loadTable` (singular) to run only once with `forceRefresh` always false. Then I added a `forceRefresh` parameter to `loadTables` (plural) that will only be true when the user clicks on the refresh button.

michael-s-molina commented:
/testenv up

villebro reviewed (approved):
LGTM, thanks for all the hard work here!
